### PR TITLE
[`pydoclint`] Permit yielding `None` in DOC402 and DOC403

### LINKED ISSUE #13001: DOC403 behaviour with yielding none
* List of keywords you searched for before creating this issue. Write them down here so that others can find this issue more easily and help provide feedback.
  "DOC403 ", "numpydoc", "yields"
* A minimal code snippet that reproduces the bug.
```
@pytest.fixture
def db_setup() -> Generator[None, None, None]:
    """
    Add test user.

    Yields
    ------
        And then cleans up test user.
    """
    database = Database('my_database_url')
    with database.create_session() as session:
        session.execute(insert(database.Users, [{'id': 1, 'name': 'Bob'}

    yield

    with database.create_session() as session:
        session.execute(delete(database.Users).where(database.Users.id == 1234)
```
* The command you invoked (e.g., `ruff /path/to/file.py --fix`), ideally including the `--isolated` flag.
`ruff /path/to/file.py --fix`
* The current Ruff settings (any relevant sections from your `pyproject.toml`).
```
[tool.ruff]
line-length = 120
preview = true
target-version = 'py312'

[tool.ruff.format]
line-ending = 'lf'
quote-style = 'single'

[tool.ruff.lint]
extend-select = ['D213']
ignore = ['ANN401', 'CPY001', 'D107', 'D203', 'D212', 'E203', 'ISC003', 'PLC1901', 'PLR6104']
select = ['ALL']

[tool.ruff.lint.flake8-implicit-str-concat]
allow-multiline = false

[tool.ruff.lint.flake8-quotes]
inline-quotes = 'single'

[tool.ruff.lint.flake8-type-checking]
strict = true

[tool.ruff.lint.pydocstyle]
convention = 'numpy'

[tool.ruff.lint.pylint]
max-args = 14
max-statements = 75

```
* The current Ruff version (`ruff --version`).
 ruff 0.6.1

The above code snippet results in a DOC403 Docstring has a "Yields" section but the function doesn't yield anything.

If the yields section is removed from the docstring then [numpydoc-validation](https://numpydoc.readthedocs.io/en/latest/validation.html) reports:

```
+---------------------+-------------------+---------+-------------------------+
| file                | item              | check   | description             |
+=====================+===================+=========+=========================+
| src/conftest.py:135 | conftest.db_setup | YD01    | No Yields section found |
+---------------------+-------------------+---------+-------------------------+
```

I understand that it is debatable about what the docstring should be a for a yield statement that yields None, but I think ruff should probably match the numpydoc-validator tool in this case.

### PR DESCRIPTION
**Pull request by tjkuson**
## Summary

Currently, [docstring-extraneous-yields (DOC403)](https://docs.astral.sh/ruff/rules/docstring-extraneous-yields/) reports a diagnostic if there is a yields section in a docstring for a function that yields `None` implicitly; for example,

```python
from collections import abc

def gen() -> abc.Generator[None, None, None]:
    """A very helpful docstring.

    Yields:
        When foo.
    """
    ...  # some code
    yield
    ...  # some more code
```

This is problematic, as sometimes you want to document this yielding behaviour.

This patch permits such docstrings by now counting implicit `yield None` statements as yields. It then updates [docstring-missing-yields (DOC402)](https://docs.astral.sh/ruff/rules/docstring-missing-yields/) to allow users to forego a yield section if the function is resolved to yield `None` only (by checking the return annotation, or by checking the yield statements if the return type is not annotated).

This is similar to #13064

Closes #13001

## Test Plan

`cargo nextest run`

### CODE REVIEW

**Comment by tjkuson** [comment]
Oh I should add `Iterator` etc to the generator types

**Comment by AlexWaygood** [comment]
So, consider this ecosystem hit: <https://github.com/apache/superset/blob/07985e2f5aa165f6868abbf88594e6d75300caae/superset/utils/core.py#L1288-L1312>. The yield is "documented" there in that it's clear that the function is a context manager that only _temporarily_ overrides a value. But there's no "Yields" section -- and I think it would make the docstring worse if they rewrote it to include a "Yields" section; but this change would mean that we'd emit DOC402 on that docstring.

I think my preferred behaviour here would be:
- Don't report `DOC402` for a function that only ever has bare `yield` or `yield None` statements
- But also, don't report `DOC403` just _because_ the `yield` is documented

I think this is what you suggested in https://github.com/astral-sh/ruff/issues/13001#issuecomment-2307604996 (sorry that nobody responded to you there!). It's a little different from numpydoc's behaviour IIUC, but I think that's okay.

**Comment by AlexWaygood** [comment]
(Oh and thanks for the PR! Sorry for diving straight into the review 😄)

**Comment by tjkuson** [comment]
> So, consider this ecosystem hit: https://github.com/apache/superset/blob/07985e2f5aa165f6868abbf88594e6d75300caae/superset/utils/core.py#L1288-L1312. 

Would you mind clarifying this and how it interacts with the current implementation of `DOC201`? From what I understand, its annotated yield type is `typing.Any`, so its yield behaviour *should* be documented (it's not `None`).

If the recommendation is to ignore the type annotation if we determine all yields to be `None` by parsing the function body, this would be inconsistent with `DOC201` behaviour of treating the return annotation as truth https://github.com/astral-sh/ruff/pull/13064. Would that implementation have to change as well, or would it just be a documented discrepancy?

**Comment by AlexWaygood** [comment]
> > So, consider this ecosystem hit: [apache/superset@`07985e2`/superset/utils/core.py#L1288-L1312](https://github.com/apache/superset/blob/07985e2f5aa165f6868abbf88594e6d75300caae/superset/utils/core.py#L1288-L1312).
> 
> Would you mind clarifying this and how it interacts with the current implementation of `DOC201`? From what I understand, its annotated yield type is `typing.Any`, so its yield behaviour _should_ be documented (it's not `None`).

Ahh, thanks for pointing that out. Your argument makes sense; I agree that it's important for these rules to remain consistent.

Your PR however creates a new inconsistency between the two rules as it currently stands, however. `DOC201` will currently complain about the first of these two docstrings, but not the second:

```py
def foo() -> int | None:
    """Foo-y method"""
    return None


def bar() -> int | None:
    """Bar-y method"""
    return
```

Would you like to fix that as well as part of this PR, so that the rules stay consistent?

**Comment by tjkuson** [comment]
> > > So, consider this ecosystem hit: [apache/superset@`07985e2`/superset/utils/core.py#L1288-L1312](https://github.com/apache/superset/blob/07985e2f5aa165f6868abbf88594e6d75300caae/superset/utils/core.py#L1288-L1312).
> > 
> > 
> > Would you mind clarifying this and how it interacts with the current implementation of `DOC201`? From what I understand, its annotated yield type is `typing.Any`, so its yield behaviour _should_ be documented (it's not `None`).
> 
> Ahh, thanks for pointing that out. Your argument makes sense; I agree that it's important for these rules to remain consistent.
> 
> Your PR however creates a new inconsistency between the two rules as it currently stands, however. `DOC201` will currently complain about the first of these two docstrings, but not the second:
> 
> ```python
> def foo() -> int | None:
>     """Foo-y method"""
>     return None
> 
> 
> def bar() -> int | None:
>     """Bar-y method"""
>     return
> ```
> 
> Would you like to fix that as well as part of this PR, so that the rules stay consistent?

Fixed in https://github.com/astral-sh/ruff/pull/13148/commits/79fad1c73c829372355d6ae8ea7f0e778f1bdfda!

**Comment by charliermarsh** [comment]
This makes sense to me, but I'll defer to @AlexWaygood to approve and merge.

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs [hunk 1]
```diff
@@ -711,6 +732,28 @@ fn yields_documented(
         || (matches!(convention, Some(Convention::Google)) && starts_with_yields(docstring))
 }
 
+/// Returns the first subscript element of a generator annotation, if it exists.
+fn generator_yield_annotation<'a>(expr: &'a Expr, checker: &'a Checker) -> Option<&'a Expr> {
+    let slice = expr.as_subscript_expr()?.slice.as_tuple_expr()?;
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:737]: Unfortunately while `Generator[int, str, None]` has an `Expr::Tuple(_)` inside the `Generator[]` slice, `Iterator[Any]` has an `Expr::Name(_)` inside the `Iterator` slice. So the logic you're adding currently works for functions returning `Generator[]` but not those returning `Iterator[]`. I'm working on a fix for this and a few more edge cases I spotted; I'll push them to your PR branch

File: crates/ruff_linter/resources/test/fixtures/pydoclint/DOC201_numpy.py [hunk 2]
```diff
@@ -187,3 +187,9 @@ def foo(s: str) -> str | None:
         A string.
     """
     return None
+
+
+# DOC201
+def bar() -> int | None:
+    """Bar-y method"""
```
[the remark this answers is not in the document]
Comment (tjkuson) [reply crates/ruff_linter/resources/test/fixtures/pydoclint/DOC201_numpy.py:194]: Can't build upon perfection

File: crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs (hunk 1 above)
Comment (tjkuson) [reply crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs:737]: Well spotted, thank you!

**Comment by AlexWaygood** [comment]
Thanks @tjkuson!
